code/tests/unittests/tests_hit_record.cpp

```diff
@@ -0,0 +1,62 @@
+#include "gtest/gtest.h"
+#include "vector3.h"
+#include "ray.h"
+#include "hitrecord.h"
+
+
+class HitRecordTest : public ::testing::Test {
+protected:
+    HitRecord record;
+    Vector3 ray_origin = Vector3(0, 0, 0);
+    Vector3 forward_direction = Vector3(0, 0, 1);
+    Ray ray;
+
+    virtual void SetUp() {
+        record = HitRecord();
+        ray = Ray(ray_origin, forward_direction);
+    }
+};
+
+TEST_F(HitRecordTest, ComputeFaceNormalDirectionIfRayHitsOutwardSetsFrontFaceToTrue) {
+    Vector3 hit_point(0, 0, .5);
+    Vector3 center(0, 0, 1);
+    Vector3 outward_normal = hit_point - center;
+
+    record.ComputeFaceNormalDirection(ray, outward_normal);
+
+    ASSERT_TRUE(record.front_face);
+}
+
+TEST_F(HitRecordTest, ComputeFaceNormalDirectionIfRayHitsOutwardSetsNormalToOutwardNormal) {
+    Vector3 hit_point(0, 0, .5);
+    Vector3 center(0, 0, 1);
+    Vector3 outward_normal = hit_point - center;
+
+    record.ComputeFaceNormalDirection(ray, outward_normal);
+
+    ASSERT_DOUBLE_EQ(record.normal.x(), outward_normal.x());
+    ASSERT_DOUBLE_EQ(record.normal.y(), outward_normal.y());
+    ASSERT_DOUBLE_EQ(record.normal.z(), outward_normal.z());
```

Comment:
I just learned that there are [matchers for arrays](https://google.github.io/googletest/gmock_cheat_sheet.html#container-matchers). How about replacing these three assertions with the following:

```
// the top of your file is here
using namespace testing;

// some more code here

ASSERT_THAT(record.normal.e, ElementsAreArray(outward_normal.e));
```

I just checked, and it has a nice report in case one of the vector's components is not as expected (I just hardcoded `123456` to provoke a failure):

```
[ RUN      ] HitRecordTest.ComputeFaceNormalDirectionIfRayHitsOutwardSetsNormalToOutwardNormal
/home/tests/unittests/tests_hit_record.cpp:43: Failure
Value of: record.normal.e
Expected: has 3 elements where
element #0 is equal to 123456,
element #1 is equal to 0,
element #2 is equal to -0.5
  Actual: { 0, 0, -0.5 } (of type double [3]), whose element #0 doesn't match
[  FAILED  ] HitRecordTest.ComputeFaceNormalDirectionIfRayHitsOutwardSetsNormalToOutwardNormal (0 ms)
```

Reply:
Done, this looks awesome!